tests/src/Simple/HelloWasm/Program.cs

```diff
@@ -950,14 +954,104 @@ private static void TestBoxToGenericTypeFromDirectMethod()
 
     public struct GenStruct<TKey>
     {
-        private TKey key; 
+        private TKey key;
 
         public GenStruct(TKey key)
         {
             this.key = key;
         }
     }
 
+    private static void TestGenericStructHandling()
+    {
+        StartTest("Casting of generic structs on return and in call params");
+
+        // test return  type is cast
+        ActualStructCallParam(new string[0]);
+
+        // test call param is cast
+        GenStructCallParam(new GenStructWithImplicitOp<string>());
+
+        // replicate compilation error with https://github.com/dotnet/corefx/blob/e99ec129cfd594d53f4390bf97d1d736cff6f860/src/System.Collections.Immutable/src/System/Collections/Immutable/SortedInt32KeyNode.cs#L561
+        new GenClassUsingFieldOfInnerStruct<GenClassWithInnerStruct<string>.GenInterfaceOverGenStructStruct>(
+            new GenClassWithInnerStruct<string>.GenInterfaceOverGenStructStruct(), null).Create();
+
+        PassTest();
+    }
+
+    public class GenClassWithInnerStruct<TKey>
+    {
+        internal readonly struct GenInterfaceOverGenStructStruct 
+        {
+            // 2 fields to avoid struct collapsing to an i32
+            private readonly TKey _firstValue;
+            private readonly TKey _otherValue;
+
+            private GenInterfaceOverGenStructStruct(TKey firstElement)
+            {
+                _firstValue = firstElement;
+                _otherValue = firstElement;
+            }
+        }
+    }
+
+    public class GenClassUsingFieldOfInnerStruct<T>
+    {
+        private readonly T _value;
+        private GenClassUsingFieldOfInnerStruct<T> _left;
+
+        public GenClassUsingFieldOfInnerStruct(T v, GenClassUsingFieldOfInnerStruct<T> left)
+        {
+            _value = v;
+            _left = left;
+        }
+
+        public GenClassUsingFieldOfInnerStruct<T> Create(GenClassUsingFieldOfInnerStruct<T> left = null)
+        {
+            // some logic to get _value in a temp 
+            return new GenClassUsingFieldOfInnerStruct<T>(_value, left ?? _left);
+        }
+    }
+
+    private static void TestGenericCallInFinally()
+    {
+        StartTest("calling generic method requiring context from finally block");
+
+        CallGenericInFinally<string>();
+
+        PassTest();
+    }
+
+    private static void CallGenericInFinally<T>()
+    {
+        List<T> list = new List<T>();
+        foreach (var s in list) // will generate a call to Dispose
```

Comment:
Would it make sense to add a test that also uses the generic context for something and we validate that it runs?

Reply:
Yes, that would be a lot better.  Have made that change resulting in the following LLVM in the test which exercises as you suggest:

```
define void @"HelloWasm_Program__CallGenericInFinally<System___Canon>$Finally5"(i8*, i8*) {
Block5:
  call void @llvm.donothing(), !dbg !15089
  %2 = getelementptr i8, i8* %0, i32 4, !dbg !15090
  %3 = getelementptr i8, i8* %0, i32 4, !dbg !15090
  %getHelper = call i8* @"__GenericLookupFromDict_HelloWasm_Program__CallGenericInFinally<System___Canon>_TypeHandle_HelloWasm_Program_GenRequiresContext_1<T_System___Canon>"(i8* %3, i8* %1), !dbg !15090
  call void @"HelloWasm_Program_GenRequiresContext_1<System___Canon>__Dispose"(i8* %2, i8* %getHelper), !dbg !15090
  call void @llvm.donothing(), !dbg !15090
  call void @llvm.donothing(), !dbg !15091
  ret void, !dbg !15091
}
```
and
```
define void @"HelloWasm_Program_GenRequiresContext_1<System___Canon>__Dispose"(i8*, i8*) {
Prolog:
  br label %Block0

Block0:                                           ; preds = %Prolog
  call void @llvm.donothing(), !dbg !15092
  %2 = getelementptr i8, i8* %0, i32 0, !dbg !15094
  %getHelper = call i8* @"__GenericLookupFromType_HelloWasm_Program_GenRequiresContext_1<System___Canon>_NonGCStaticBase_HelloWasm_Program_GenRequiresContext_1<T_System___Canon>"(i8* %2, i8* %1), !dbg !15094
  %Called_addr = getelementptr i8, i8* %getHelper, i32 0, !dbg !15094
  %CastPtrbool = bitcast i8* %Called_addr to i1*, !dbg !15094
  store i1 true, i1* %CastPtrbool, !dbg !15094
  ret void, !dbg !15095
}
```